### PR TITLE
Unnest fixing and re-address validation triggers

### PR DESCRIPTION
**Pull request by alanmcruickshank**
Found this while working on #5230.

The `.apply_fixes()` method was very deeply nested, in a way that made the code hard to follow. This flips some `if` statments so that we can un-nest that code, make it easier to follow and cleaner.

In the process I've also slightly changed the conditions for validating things. The the segment _type_ is the same and we're just editing the content (e.g. capitalisation changes) then we don't do a full re-parse validation (and we trust that the rule is being sensible). For anything else more significant we do still trigger a validation though.

### CODE REVIEW

File: src/sqlfluff/core/parser/segments/base.py [hunk 1]
```diff
@@ -1219,79 +1219,90 @@ def apply_fixes(
...
         fixes_applied: List[LintFix] = []
         todo_buffer = list(self.segments)
-        while True:
-            if len(todo_buffer) == 0:
-                break
-            else:
-                seg = todo_buffer.pop(0)
+        requires_validate = False
+
+        while todo_buffer:
```
Comment (alanmcruickshank) [inline src/sqlfluff/core/parser/segments/base.py:1224]: rather than `while True: ... if todo_buffer...`, better to just have `while todo_buffer:`

File: src/sqlfluff/core/parser/segments/base.py [hunk 2]
```diff
@@ -1219,79 +1219,90 @@ def apply_fixes(
...
-                # Look for uuid match.
-                # This handles potential positioning ambiguity.
-                anchor_info: Optional[AnchorEditInfo] = fixes.pop(seg.uuid, None)
-                if anchor_info is not None:
-                    seg_fixes = anchor_info.fixes
+            # Look for uuid match.
+            # This handles potential positioning ambiguity.
+            anchor_info: Optional[AnchorEditInfo] = fixes.pop(seg.uuid, None)
+
+            if anchor_info is None:
```
Comment (alanmcruickshank) [inline src/sqlfluff/core/parser/segments/base.py:1231]: Flip this `if` to deal with the `is None` case first and then `continue`, followed by an un nested opposite case.

File: src/sqlfluff/core/parser/segments/base.py [hunk 3]
```diff
@@ -1219,79 +1219,90 @@ def apply_fixes(
...
+                    requires_validate = True
+                    # NOTE: We don't add the segment in this case.
+                    continue
+
+                # Otherwise it must be a replace or a create.
+                assert f.edit_type in (
+                    "replace",
+                    "create_before",
+                    "create_after",
+                ), f"Unexpected edit_type: {f.edit_type!r} in {f!r}"
```
Comment (alanmcruickshank) [inline src/sqlfluff/core/parser/segments/base.py:1267]: Changed this `if` to an `assert` and then removed the `raise` clause later.

**Review by greg-finley — APPROVED** [review]
